### ISSUE TITLE
Customer Churn Prediction Using Machine Learning

### ISSUE THREAD

**Issue by mannam-30** [mannam-30/mannam.github.io#2]
# Import necessary libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, roc_auc_score
import shap

# Load the dataset
df = pd.read_csv('Telco-Customer-Churn.csv')

# Data preprocessing
df = df.dropna()  # Handle missing values

# Encode categorical features
categorical_features = ['gender', 'Partner', 'Dependents', 'PhoneService', 
                        'MultipleLines', 'InternetService', 'OnlineSecurity', 
                        'OnlineBackup', 'DeviceProtection', 'TechSupport', 
                        'StreamingTV', 'StreamingMovies', 'Contract', 
                        'PaperlessBilling', 'PaymentMethod']

# Feature and label split
X = df.drop(columns=['customerID', 'Churn'])
y = df['Churn'].apply(lambda x: 1 if x == 'Yes' else 0)

# One-hot encode categorical variables
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), ['tenure', 'MonthlyCharges', 'TotalCharges']),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_features)
    ])

# Model pipeline
model = Pipeline(steps=[('preprocessor', preprocessor),
                        ('classifier', RandomForestClassifier(n_estimators=100, random_state=42))])

# Split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Train the model
model.fit(X_train, y_train)

# Predict on the test set
y_pred = model.predict(X_test)

# Evaluate the model
print(classification_report(y_test, y_pred))
print('ROC-AUC Score:', roc_auc_score(y_test, model.predict_proba(X_test)[:, 1]))

# Model explanation with SHAP
explainer = shap.TreeExplainer(model.named_steps['classifier'])
shap_values = explainer.shap_values(model.named_steps['preprocessor'].transform(X_test))
shap.summary_plot(shap_values[1], model.named_steps['preprocessor'].transform(X_test), feature_names=X.columns)
